Source/JavaScriptCore/heap/MarkedBlock.cpp

```diff
@@ -502,6 +506,67 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
     specializedSweep<false, IsEmpty, SweepOnly, BlockHasNoDestructors, DontScribble, HasNewlyAllocated, MarksStale>(freeList, emptyMode, sweepMode, BlockHasNoDestructors, scribbleMode, newlyAllocatedMode, marksMode, [] (VM&, JSCell*) { });
 }
 
+NO_RETURN_DUE_TO_CRASH NEVER_INLINE void MarkedBlock::dumpInfoIfHandleIsNotValid(AbstractLocker&, JSCell* cell)
+{
+    size_t contiguousZeroCount = 0;
+    {
+        char* mem = WTF::bitwise_cast<char*>(&header());
+        for (; contiguousZeroCount < sizeof(MarkedBlock::blockSize); ++contiguousZeroCount) {
+            if (*mem)
+                break;
+            ++mem;
+        }
+    }
+
+    bool isValidBlockVM = false;
+    bool isBlockInVM = false;
+    bool isBlockHandleInVM = false;
+    VM* blockVM = header().m_vm;
+    {
+        isValidBlockVM = blockVM ? VMInspector::isValidVM(blockVM) : false;
+        if (isValidBlockVM) {
+            Heap& heap = blockVM->heap;
+            MarkedSpace& objectSpace = heap.objectSpace();
+            isBlockInVM = objectSpace.blocks().set().contains(this);
+            isBlockHandleInVM = !!objectSpace.findMarkedBlockHandle(this);
+        } else {
+            VMInspector::forEachVM([&](VM& vm) {
+                MarkedSpace& objectSpace = vm.heap.objectSpace();
+                isBlockInVM = objectSpace.blocks().set().contains(this);
+                isBlockHandleInVM = !!objectSpace.findMarkedBlockHandle(this);
+                // Either of them is true indicates that the block belongs or used to belong to the VM.
+                if (isBlockInVM || isBlockHandleInVM) {
+                    blockVM = &vm;
+                    return IterationStatus::Done;
+                }
+                return IterationStatus::Continue;
+            });
+        }
+    }
+
+    bool hasCell = !!cell;
+    bool isValidCellVM = false;
+    bool isCellVMSameAsBlockVM = false;
+    bool isAliveCell = false;
+    if (hasCell) {
+        VM& cellVM = cell->vm();
+        isValidCellVM = VMInspector::isValidVM(&cellVM);
+        isCellVMSameAsBlockVM = blockVM == &cellVM;
+        if (isValidCellVM)
+            isAliveCell = VMInspector::isValidCell(&cellVM.heap, cell);
+    }
```

Comment:
This isn't going to do anything since `HeapCell::vm()` just masks the bits off to get to the `MarkedBlock` and returns the `MarkedBlock`'s vm. Probably, the only meaningful parts of the cell, if it is one, is the cell's JSType and what offset into the MarkedBlock it is.

Reply:
Rather than just the offset, it might be worth recording the entire address of the cell. We may find that it points into some region that doesn't make sense (but is mapped readable), meaning it wasn't the block header that was corrupted but that we chased an invalid pointer already.

Reply:
I should dump the addresses of the `VM` and `JSCell`. And the offset is just the a subtraction.